subprojects/dependency-management/src/main/java/org/gradle/api/internal/artifacts/ivyservice/resolveengine/graph/builder/ModuleSelectors.java

```diff
@@ -45,147 +67,85 @@ public boolean checkDeferSelection() {
     @SuppressWarnings("unchecked")
     @Override
     public Iterator<T> iterator() {
-        if (selectorsCount == 0) {
-            return EMPTY_ITERATOR;
-        } else if (selectorsCount == 1) {
-            return Iterators.singletonIterator(singleSelector);
-        } else {
-            if (selectors != null && dynamicSelectors != null) {
-                return Iterators.concat(selectors.iterator(), dynamicSelectors.iterator());
-            } else if (selectors != null) {
-                return selectors.iterator();
-            } else {
-                return dynamicSelectors.iterator();
-            }
+        sort();
+        return selectors.iterator();
+    }
+
+    private void sort() {
+        if (shouldSort) {
+            Collections.sort(selectors, SELECTOR_COMPARATOR);
+            shouldSort = false;
         }
     }
 
     public void add(T selector, boolean deferSelection) {
-        assert !contains(selector) : "Inconsistent call to add: should only be done if the selector isn't in use";
-        if (selectorsCount == 0) {
-            singleSelector = selector;
-            this.deferSelection = deferSelection;
-        } else if (selectorsCount == 1) {
-            addSelector(singleSelector);
-            addSelector(selector);
-            singleSelector = null;
-        } else {
-            addSelector(selector);
-        }
-        selectorsCount++;
+        this.deferSelection = deferSelection;
+        selectors.add(selector);
+        size++;
+        shouldSort = size > 1;
     }
 
-    private void addSelector(T selector) {
-        if (isDynamicSelector(selector)) {
-            addDynamicSelector(selector);
-        } else {
-            addSimpleSelector(selector);
+    public boolean remove(T selector) {
+        boolean remove = selectors.remove(selector);
+        if (remove) {
+            size--;
+            shouldSort = size > 1;
```

Comment:
There should not be a need to re-sort when removing something, as removal preserves order.

Reply:
Good catch!